File: backend/app/services/access_score.py
"""
Access Score computation — straight-line distances + density scoring.
Upgrade path: swap haversine for cached drive-time lookups.
"""
import math
from dataclasses import dataclass


FACILITY_WEIGHTS = {
    "elevator": 0.35,
    "ethanol": 0.20,
    "processor": 0.20,
    "rail": 0.15,
    "river": 0.10,
}

# Distance thresholds (miles) for scoring
DIST_THRESHOLDS = {"excellent": 10, "good": 25, "fair": 50, "poor": 100}


def haversine_miles(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    R = 3958.8  # Earth radius in miles
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (math.sin(dlat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(dlon / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


@dataclass
class FacilityResult:
    type: str
    nearest_distance: float | None
    count_25mi: int
    count_50mi: int
    sub_score: float  # 0-100


def score_distance(dist: float | None) -> float:
    if dist is None:
        return 0
    if dist <= DIST_THRESHOLDS["excellent"]:
        return 100
    if dist <= DIST_THRESHOLDS["good"]:
        return 80 - (dist - 10) * (30 / 15)
    if dist <= DIST_THRESHOLDS["fair"]:
        return 50 - (dist - 25) * (25 / 25)
    if dist <= DIST_THRESHOLDS["poor"]:
        return 25 - (dist - 50) * (20 / 50)
    return max(0, 5 - (dist - 100) * 0.05)


def score_density(count_25: int, count_50: int) -> float:
    """Bonus for facility density (0-20 points)."""
    return min(20, count_25 * 5 + count_50 * 1.5)


def compute_access_score(
    county_lat: float,
    county_lon: float,
    facilities: list[dict],  # [{type, lat, lon, name, ...}]
    weights: dict[str, float] | None = None,
) -> dict:
    """Compute the 0-100 access score for a county centroid."""
    w = weights or FACILITY_WEIGHTS

    by_type: dict[str, list[float]] = {}
    for f in facilities:
        t = f["type"]
        d = haversine_miles(county_lat, county_lon, f["lat"], f["lon"])
        by_type.setdefault(t, []).append(d)

    results: dict[str, FacilityResult] = {}
    total_score = 0.0
    total_weight = 0.0

    for ftype, weight in w.items():
        distances = sorted(by_type.get(ftype, []))
        nearest = distances[0] if distances else None
        count_25 = sum(1 for d in distances if d <= 25)
        count_50 = sum(1 for d in distances if d <= 50)

        dist_score = score_distance(nearest)
        dens_bonus = score_density(count_25, count_50)
        sub = min(100, dist_score * 0.8 + dens_bonus)

        results[ftype] = FacilityResult(
            type=ftype,
            nearest_distance=round(nearest, 1) if nearest else None,
            count_25mi=count_25,
            count_50mi=count_50,
            sub_score=round(sub, 1),
        )
        total_score += sub * weight
        total_weight += weight

    composite = round(total_score / total_weight, 1) if total_weight > 0 else 0

    return {
        "access_score": composite,
        "details": {k: vars(v) for k, v in results.items()},
        "distances_json": {
            k: v.nearest_distance for k, v in results.items()
        },
        "density_json": {
            f"{k}_25mi": v.count_25mi for k, v in results.items()
        } | {
            f"{k}_50mi": v.count_50mi for k, v in results.items()
        },
    }

File: deploy/cloudflare-worker/src/services/access-score.ts
/**
 * Access Score Service — facility proximity and density scoring.
 * Ported from Python backend/app/services/access_score.py
 */

export interface Facility {
  id: number;
  type: string;
  name: string;
  lat: number;
  lon: number;
}

export interface AccessResult {
  access_score: number;
  details: Record<string, FacilityTypeDetail>;
  distances_json: Record<string, number>;
  density_json: Record<string, number>;
}

interface FacilityTypeDetail {
  type: string;
  nearest_distance: number;
  count_25mi: number;
  count_50mi: number;
  sub_score: number;
}

// Facility type weights (must sum to 1.0)
const WEIGHTS: Record<string, number> = {
  elevator: 0.35,
  ethanol: 0.20,
  processor: 0.20,
  rail: 0.15,
  river: 0.10,
};

/**
 * Haversine distance between two lat/lon points in miles.
 */
export function haversine(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 3958.8; // Earth radius in miles
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

/**
 * Distance-based score (0-100) for a single facility.
 */
function distanceScore(miles: number): number {
  if (miles <= 10) return 100;
  if (miles <= 25) return 80 - ((miles - 10) / 15) * 30; // 80→50
  if (miles <= 50) return 50 - ((miles - 25) / 25) * 25; // 50→25
  if (miles <= 100) return 25 - ((miles - 50) / 50) * 20; // 25→5
  return Math.max(0, 5 * Math.exp(-(miles - 100) / 100));
}

/**
 * Density bonus (0-20 points) based on number of facilities within radius.
 */
function densityBonus(count25: number, count50: number): number {
  return Math.min(20, count25 * 5 + count50 * 2);
}

/**
 * Compute access score for a county given its centroid and all facilities.
 */
export function computeAccessScore(
  lat: number,
  lon: number,
  facilities: Facility[],
): AccessResult {
  const typeGroups: Record<string, { distance: number; count25: number; count50: number }[]> = {};

  // Group facilities by type and compute distances
  for (const f of facilities) {
    const dist = haversine(lat, lon, f.lat, f.lon);
    const type = f.type;
    if (!typeGroups[type]) typeGroups[type] = [];
    typeGroups[type].push({
      distance: dist,
      count25: dist <= 25 ? 1 : 0,
      count50: dist <= 50 ? 1 : 0,
    });
  }

  const details: Record<string, FacilityTypeDetail> = {};
  const distances: Record<string, number> = {};
  const density: Record<string, number> = {};
  let weightedScore = 0;
  let totalWeight = 0;

  for (const [type, weight] of Object.entries(WEIGHTS)) {
    const group = typeGroups[type] || [];
    if (group.length === 0) {
      details[type] = { type, nearest_distance: 999, count_25mi: 0, count_50mi: 0, sub_score: 0 };
      distances[type] = 999;
      density[`${type}_25mi`] = 0;
      density[`${type}_50mi`] = 0;
      continue;
    }

    const nearest = Math.min(...group.map((g) => g.distance));
    const count25 = group.reduce((s, g) => s + g.count25, 0);
    const count50 = group.reduce((s, g) => s + g.count50, 0);
    const subScore = Math.min(100, distanceScore(nearest) + densityBonus(count25, count50));

    details[type] = { type, nearest_distance: round1(nearest), count_25mi: count25, count_50mi: count50, sub_score: round1(subScore) };
    distances[type] = round1(nearest);
    density[`${type}_25mi`] = count25;
    density[`${type}_50mi`] = count50;

    weightedScore += subScore * weight;
    totalWeight += weight;
  }

  const finalScore = totalWeight > 0 ? round1(weightedScore / totalWeight) : 0;

  return {
    access_score: finalScore,
    details,
    distances_json: distances,
    density_json: density,
  };
}

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

function round1(n: number): number {
  return Math.round(n * 10) / 10;
}
